vote/src/vote_account.rs

```diff
@@ -182,37 +212,61 @@ impl VoteAccounts {
     }
 
     fn add_node_stake(&mut self, stake: u64, vote_account: &VoteAccount) {
-        if stake == 0u64 {
-            return;
-        }
         let Some(staked_nodes) = self.staked_nodes.get_mut() else {
             return;
         };
-        if let Some(node_pubkey) = vote_account.node_pubkey() {
-            Arc::make_mut(staked_nodes)
-                .entry(node_pubkey)
-                .and_modify(|s| *s += stake)
-                .or_insert(stake);
-        }
+
+        VoteAccounts::do_add_node_stake(staked_nodes, stake, vote_account.node_pubkey());
     }
 
-    fn sub_node_stake(&mut self, stake: u64, vote_account: &VoteAccount) {
+    fn do_add_node_stake(
+        staked_nodes: &mut Arc<HashMap<Pubkey, u64>>,
+        stake: u64,
+        node_pubkey: Option<&Pubkey>,
+    ) {
         if stake == 0u64 {
             return;
         }
+
+        node_pubkey.map(|node_pubkey| {
+            Arc::make_mut(staked_nodes)
+                // Note we take an Option<&Pubkey> instead of taking by value, which is usually a
+                // bit of an anti pattern, but none of the callers of this function have an owned
+                // Option<Pubkey> so effectively this avoids a copy when node_pubkey is None.
+                .entry(*node_pubkey)
+                .and_modify(|s| *s += stake)
+                .or_insert(stake)
+        });
+    }
+
+    fn sub_node_stake(&mut self, stake: u64, vote_account: &VoteAccount) {
         let Some(staked_nodes) = self.staked_nodes.get_mut() else {
             return;
         };
-        if let Some(node_pubkey) = vote_account.node_pubkey() {
-            let Entry::Occupied(mut entry) = Arc::make_mut(staked_nodes).entry(node_pubkey) else {
-                panic!("this should not happen!");
-            };
-            match entry.get().cmp(&stake) {
+
+        VoteAccounts::do_sub_node_stake(staked_nodes, stake, vote_account.node_pubkey());
+    }
+
+    fn do_sub_node_stake(
+        staked_nodes: &mut Arc<HashMap<Pubkey, u64>>,
+        stake: u64,
+        node_pubkey: Option<&Pubkey>,
+    ) {
+        if stake == 0u64 {
+            return;
+        }
+
+        if let Some(node_pubkey) = node_pubkey {
+            let staked_nodes = Arc::make_mut(staked_nodes);
+            let current_stake = staked_nodes
+                .get_mut(node_pubkey)
+                .expect("this should not happen");
```

Comment:
Just curious, why not use the entry api here?

Reply:
It's a micro opt since HashMap::entry takes the key by value in case it needs to insert. We never insert here, and the case in which we remove is so uncommon that it's worth re-hashing in that case rather than copying the key in every case.

In most cases this kind of optimization doesn't really matter since Pubkey is only 32 bytes, so copy isn't slow. Since we're in a critical section holding a write lock here tho, I think it makes sense to try and make things as fast as possible.